### ISSUE TITLE
Are setter names in the scope of an imported or exported library?

### ISSUE THREAD

**Issue by leafpetersen** [dart-lang/language#1471]
The following code is considered erroneous by both the analyzer and the CFE, despite there being a setter named `x` declared in `lib1.dart`.  The spec implies that this introduces a name `x=` into the scope, so I would expect that such a name could be shown/hidden.  Is this WAI, or is this a bug in the tooling?

```dart
import "lib1.dart" show x=;

export "lib1.dart" show x=;
```

cc @eernstg @munificent @lrhn

**Comment by eernstg** [comment]
I'd say it is a grammar issue: We can have `'show' <identifierList>` as part of an import directive, and this doesn't allow for `x=`. Access to getters and setters is in general governed by lookups for the basename, so we'd never look up `x=`, and an import directive can't show or hide a setter without doing the same thing to the corresponding getter (if any).

I think the only situation where the `=` is actually present is `noSuchMethod` symbols:

```dart
class C {
  noSuchMethod(Invocation invocation) {
    print(invocation.memberName);
    return null;
  }
}

main() {
  dynamic d = C();
  d.setter = 42; // Prints 'Symbol("setter=")'.
}
```

**Comment by lrhn** [comment]
Agree, the `show`/`hide` for that setter would be `show x`. That's well in line with out move towards treating properties as a single thing, not unrelated getters and setters. The grammar doesn't allow you to write `x=` as a symbol anywhere, not even in `#x=` (but it should: #301).

If we want the `show`/`hide` for extension types, we might need to add `operator <op>` as well, but I'd still use `x` to refer to the property. If we want to be able to show only the getter or setter, I'd propose something like `show get x, set y`, but it's probably ambiguous since things can be named `set`, `get`, `show` and `hide`, so `show get hide show hide get` can be parse two ways ☹️ .

**Comment by leafpetersen** [comment]
> If we want the `show`/`hide` for extension types, we might need to add `operator <op>` as well, but I'd still use `x` to refer to the property.

This is why I brought it up, and it feels a bit off to me to disallow it.  I guess the argument would be that you can always choose to hide both and then manually delegate the one you want?  That feels pretty bureaucratic though.

**Comment by eernstg** [comment]
The proposal in https://github.com/dart-lang/language/pull/1452/files already has this, and that doesn't create any parsing issues:

```ebnf
<extensionShowHideElement> ::= <type> | <identifier> | 'operator' <operator>
```

With respect to the getter/setter bundle, it is indeed somewhat inconvenient that we can't immediately use show/hide to show a read-only interface as a subset of an interface with mutators, because that would surely involve handling instance variables where we want to enable the getter, but not the setter.

However, we could keep the rule that `x` refers to both the getter and setter with basename `x`, and we could use `get x` and `set x` to denote the getter/setter on its own (as mentioned by @lrhn).

```dart
class IntBox {
  int i;
  IntBox(this.i);
}

extension type ReadOnlyIntBox on IntBox show get i {}
```

I've adjusted [`Dart.g`](https://github.com/dart-lang/sdk/blob/master/tools/spec_parser/Dart.g) to allow `('get'|'set') <identifier>` in the show/hide clause of an extension, and it does not create any parsing problems.

**Comment by lrhn** [comment]
ACK, my attempt to make an ambiguous sequence failed, it wasn't actually ambiguous, and it's quite possible that it can't be. 

Your grammar is helped by only allowing one show and one hide, not an arbitrary number like class declarations. If you loosen that, parsing might needs some backtracking/look-ahead to parse, though:.
Take `show get hide hide set show`. It's not until seeing the `set` that you know that it must be `(show (get hide)(hide (set show))` instead of `(show get)(hide hide)(...)`). You *can* name members `get`, `set`, `show` and `hide`.
So, for maximum required lookahead: `show get hide hide hide hide hide hide hide hide .... hide` followed by either `x` or not.

**Comment by eernstg** [comment]
Do you have a compelling example where we'd want multiple hide clauses? I seems like it could make sense to build a set as a union of several sets, and that could be written as `show M1 hide M2 show M3 hide M4 ...` which could mean `UNION(M1\M2, M3\M4 ...)`, but I remember taking a look at the import directives and concluding that the semantics is useless for multi-hide and multi-show forms, so we are unlikely to want to copy that.

**Comment by lrhn** [comment]
If we allow showing *interfaces* then that changes because you can then show/hide multiple things in one operation, and then might you want to make *exceptions* to that. With only individual names, that doesn't show up.
I might want to do:
```dart
... on int hide int show num hide Comparable<num>, roundToDouble
```
For taller class hierarchies, you can make longer chains.
